### PR TITLE
Validate kubeconfig files in case of external CA mode

### LINKED ISSUE #1203: Missing validation of kubeconfig files in case of external CA mode
As per comment https://github.com/kubernetes/kubernetes/pull/70098#discussion_r229320061

## What you expected to happen?

External CA mode requires that the uses provides all the certs - except certificate authorithy keys - and all the kubeconfig files.

However, the check for External CA mode validates certs (existance and validity) but does not validates kubeconfig files (nether existance nor validity), and this can lead to errors.

Instead, external CA mode test should check for kubeconfig files (existance and validity) too.

## Anything else we need to know?

A possible solution is to trigger the check as part of the external CA mode test [here](https://github.com/kubernetes/kubernetes/blob/50dffba0f3f21580c0ce433096a5e54575341224/cmd/kubeadm/app/phases/certs/certs.go#L315) but the most of the implementation should be done into [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go) reusing the validation logic already implemented [here](https://github.com/kubernetes/kubernetes/blob/50dffba0f3f21580c0ce433096a5e54575341224/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L237-L255)

### PR DESCRIPTION
**Pull request by yagonobre**
**What type of PR is this?**
/kind bug
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/1203

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: validate kubeconfig files in case of external CA mode.
```

/assign @timothysc
/assign @fabriziopandini

### CODE REVIEW

**Comment by neolit123** [comment]
@yagonobre thanks.
please use this format for the release note:
```
kubeadm: validate kubeconfig files in case of external CA mode.
```

**Comment by neolit123** [comment]
/priority important-longterm
(?)

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 1]
```diff
@@ -254,6 +244,31 @@ func createKubeConfigFileIfNotExists(outDir, filename string, config *clientcmda
...
+// If there already is a kubeconfig file at the given path; kubeadm tries to load it and check if the values in the
+// existing and the expected config equals. If they do; kubeadm will just skip writing the file as it's up-to-date,
+// but if a file exists but has old content or isn't a kubeconfig file, this function returns an error.
+func createKubeConfigFileIfNotExists(outDir, filename string, config *clientcmdapi.Config) error {
+	kubeConfigFilePath := filepath.Join(outDir, filename)
+
+	err := ValidateKubeConfig(outDir, filename, config)
+	if err != nil {
+		// Check if the file exist, and if it doesn't, just write it to disk
+		if os.IsNotExist(err) {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: can be done as follows, optionally:

```
if !os.IsNotExist(err) {
	return err
}
...
```

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 2]
```diff
@@ -254,6 +244,31 @@ func createKubeConfigFileIfNotExists(outDir, filename string, config *clientcmda
...
+			fmt.Printf("[kubeconfig] Writing %q kubeconfig file\n", filename)
+
+			err = kubeconfigutil.WriteToDisk(kubeConfigFilePath, config)
+			if err != nil {
+				return errors.Wrapf(err, "failed to save kubeconfig file %s on disk", kubeConfigFilePath)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: "%s" -> "%q"

**Review by neolit123 — COMMENTED** [review]
LGTM on a quick look.
added two minor nits.

**Comment by yagonobre** [comment]
@neolit123 done

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 3]
```diff
@@ -116,7 +125,14 @@ func runKubeConfigFile(kubeConfigFileName string) func(workflow.RunData) error {
 
 		// if external CA mode, skip certificate authority generation
 		if data.ExternalCA() {
-			//TODO: implement validation of existing kubeconfig files
+			config, err := getKubeConfig(data.Cfg(), kubeConfigFileName)
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Is this new behavior covered either directly or indirectly in the _tests file?  

/assign @fabriziopandini

Comment (yagonobre) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Not yet, I'll update the tests tomorrow.

**Review by fabriziopandini — COMMENTED** [review]
@yagonobre thanks for this PR, this is definitely a step in the right direction!

However, IMO we should change where the validation is triggered in order to fail as soon as possible in case we are in external CA mode (certificate authority certs existing without keys), but without all proper certificates and kubeconfig file in places.

In order to do so, I think that the kubeconfig validation should happen before, ideally when building ìnitData`/inside the`externalCA` function [here](https://github.com/kubernetes/kubernetes/blob/808557e468639d8fd6027c99d004bceb8e8a5ec1/cmd/kubeadm/app/cmd/init.go#L317). wdyt?

@timothysc @neolit123 opinions?

**Comment by neolit123** [comment]
> In order to do so, I think that the kubeconfig validation should happen before, ideally when building ìnitData/inside theexternalCA` function here. wdyt?

+1

**Comment by neolit123** [comment]
@fabriziopandini 
@yagonobre reported in slack that golint is not happy with kubeconfig.KubeConfigSpec it should be called kubeconfig.Spec. the alternative is to use hack/.golint_failures.

options:
1. try SpecKubeConfig
2. move to Spec
3. modify hack/.golint_failures

my vote is to try 1, then 2.

**Comment by fabriziopandini** [comment]
@neolit123 @yagonobre IMO `KubeConfigSpec` struct should not be changed to public, but instead, a new `func ValidateKubeconfigsForExternalCA` should be exposed and then invoked by `ExternalCA`
(my comments above/on the issue are on the same line too)

**Comment by yagonobre** [comment]
@fabriziopandini make senses, I'll update soon

**Review by fabriziopandini — APPROVED** [review]
@yagonobre thanks for this update to this PR!
IMO this is ready to merge as soon as the new `ValidateKubeconfigsForExternalCA` func is covered by tests as required by @timothysc
/approve

**Comment by yagonobre** [comment]
Updated with some tests

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**